Texturing

timoffex opened the pull request:
Significant changes to rendering pipeline: most rendering code moved outside of MeshView into subtypes of AbstractRenderer. Currently, only the SimpleTexturedRenderer is provided.

This PR also includes the 1x1 map fix.

timoffex committed:
Merge branch 'master' into map2mesh_improvements

timoffex committed:
Unfinished MeshMaterial changes (for textures)

timoffex committed:
Merge branch 'master' into map2mesh_improvements

timoffex committed:
Merge branch 'master' into map2mesh_improvements

timoffex committed:
Created Triplet class

timoffex committed:
Added changed() signal to camera

timoffex committed:
Created DereferencingIterator class

This is a wrapper for arbitrary iterators that dereferences their
represented value.

For example, it allows one to create an iterator over a
QVector<QSharedPointer<MyType>> such that dereferencing the iterator
returns MyType& (or const MyType&) instead of QSharedPointer<MyType>.

timoffex committed:
AbstractRenderer and AbstractScene

An AbstractRenderer renders a 3D scene using OpenGL.

An AbstractScene represents a 3D scene and has an associated Renderer.
Each subtype of AbstractScene implements the makeRenderer() function to
return the appropriate subtype of AbstractRenderer.

timoffex committed:
SimpleTextured* classes

The Scene, Object and Renderer for simple textured rendering. “Simple
textured rendering” means each object has exactly one texture.

SimpleTexturedRenderer needs to own its SimpleTexturedScene, but
SimpleTexturedScene needs to be the only class that knows about
SimpleTexturedRenderer, which is why std::shared_ptr and
std::enable_shared_from_this are used (QSharedPointers don’t have an
equivalent feature).

The Object (SimpleTexturedObject) contains vertex, lighting and
texturing information. Normals are stored per-face rather than
per-vertex, as this makes more sense. The Object has set*Info() methods
for setting the data, and a commit() method that must be called when
data is finished changing.

The Scene keeps a list of Objects, which can be added or removed. It
also creates and connects to a Renderer. The Scene’s commitChanges()
method needs to be called to update its Renderer.

The Renderer currently ignores object texture information, and renders
using ShaderProgramOneLight.

timoffex committed:
Updated .pro file

This includes the classes introduced in the previous commit in the
project.

timoffex committed:
Replaced old rendering system with new one

Removed Editor::makeNewScene() (the Editor no longer has to handle how
Map2Mesh connects to MeshView!)

Map2Mesh now uses SimpleTextured* classes.

timoffex committed:
Cleaned up old files

Removed RenderableObject, Scene, MeshMaterial.

Cleaned up objtools.h and implemented it for SimpleTexturedObject. Not
yet tested!

timoffex committed:
AbstractRenderer::checkGLErrors()

This is just a helper method.

I also added an extra newline somewhere in the file.

timoffex committed:
Added SimpleTexturedShader

Also changed onelighttexture.fsh to use a directed light.

timoffex committed:
Added destroy() method to SimpleTexturedShader

It is necessary to be able to explicitly clean resources (in my future
commit).

timoffex committed:
Added textures resource file with an example texture

timoffex committed:
Changed SimpleTexturedRenderer to use SimpleTexturedShader

Also added a mutex for modifying or using data.

Rendering order is now determined by texture to minimize the number of
texture calls.

Added extra functions for dealing with textures.

SimpleTexturedObject now uses QImages instead of QOpenGLTextures. This
is because QOpenGLTextures need to be created in an OpenGL context, so
the natural place to create them is inside a renderer’s initializeGL().

M2MTileMesher now outputs a textured mesh.

Added cleanUp() to MeshView to free resources when context is changed.

timoffex committed:
Merge branch 'master' into map2mesh_improvements

# Conflicts:
#	WallsAndHoles/WallsAndHoles.pro
#	WallsAndHoles/editor.cpp
#	WallsAndHoles/map2mesh.cpp
#	WallsAndHoles/map2mesh.h

Bdtrotte commented on WallsAndHoles/WallsAndHoles.pro.orig:
```diff
@@ -0,0 +1,152 @@
+#-------------------------------------------------
```
Stray file!

Bdtrotte commented on WallsAndHoles/abstractmeshviewcamera.h:
```diff
@@ -1,17 +1,21 @@
  * Abstract class that will determine how the camera moves based on the mouse.
  */
-class AbstractMeshViewCamera : public AbstractTool {
+class AbstractMeshViewCamera : public QObject, public AbstractTool {
```
Any good reason to not just have AbstractTool inherit QObject?

timoffex replied:
Can't think of any signals that an AbstractTool would have. Maybe deactivate() and activate() could be slots, but they are probably only ever going to be called directly from a ToolManager. And (I think) mouse events in QWidgets don't have corresponding signals, so making those functions into slots wouldn't make any code simpler.

Bdtrotte replied:
But, it would allow other tools to use signals and slots, and as far as Im aware, there isn't much overhead for making it inherit QObject

Bdtrotte commented on WallsAndHoles/abstractrenderer.h:
```diff
@@ -0,0 +1,164 @@
+     */
+    void checkGLErrors()
+    {
+        GLenum error = GL_NO_ERROR;
```
doing
```
while ((GLenum error = glGetError()) != GL_NO_ERROR) { 
    switch (error) {
         ... 
    }
 }
```
could be cleaner

timoffex replied:
Will change, although GLenum error still has to be declared outside of the while loop.

timoffex committed:
Made MeshView inherit QOpenGLFunctions again

for glViewport() call

Bdtrotte commented on WallsAndHoles/simpletexturedrenderer.cpp:
```diff
@@ -0,0 +1,401 @@
+    }
+
+    // Record that the object uses the texture.
+    mTexturesToObjects[mImagesToTextures[&obj.getImage()].data()].insert(&obj);
```
Encapsulation! Maybe store the texture above the if (!mImages...) so you don't have to access the map again (And make it cleaner)

timoffex replied:
I can make it cleaner, but I am not sure what you mean by encapsulation here.

Bdtrotte replied:
I think it's one of the buzz words from class, maybe I'm miss using it...

Bdtrotte commented on WallsAndHoles/simpletexturedrenderer.cpp:
```diff
@@ -0,0 +1,401 @@
+void SimpleTexturedRenderer::paint(QMatrix4x4 mvpMatrix, QVector3D camPos)
+{
+    QMutexLocker locker(&mGLDataMutex);
+    glEnable(GL_DEPTH_TEST);
```
glEnable(GL_CULL_FACE);

Bdtrotte commented on WallsAndHoles/simpletexturedrenderer.cpp:
```diff
@@ -0,0 +1,401 @@
+    mShaderProgram.setUniformSourceSpecularColor(QVector3D(0.3, 0.3, 0.2));
+
+    // Draw objects by texture group.
+    foreach (const QImage *img, mImagesToTextures.keys()) {
```
why not `Texture t : mTexturesToObjects.keys()` ?

timoffex replied:
https://www.dvratil.cz/2015/06/qt-containers-and-c11-range-based-loops/
because it might be inefficient. Then again, according to https://www.kdab.com/goodbye-q_foreach/ it may be getting deprecated.

timoffex committed:
Merge branch 'master' into map2mesh_improvements

timoffex committed:
Removed stray .orig file

Bdtrotte commented on WallsAndHoles/simpletexturedrenderer.h:
```diff
@@ -0,0 +1,136 @@
+     * @brief Goes through the QMaps whose keys are SimpleTexturedObjects and gets rid of
+     * objects that no longer exist. Assumes the correct OpenGL context is bound. Locks the mGLDataMutex.
+     */
+    void cleanMaps();
```
I don't think this function should exist. Shouldn't it be updated when an object is removed?

timoffex replied:
Yes, it should be updated. I wrote this function just in case, before I had a signals/slots system that worked. I will remove it.

Bdtrotte commented on WallsAndHoles/editor.cpp.orig:
```diff
@@ -0,0 +1,234 @@
+#include "editor.h"
```
Stray file

timoffex replied:
They get created when I use git mergetool, and GitHub Desktop automatically adds them. Sometimes I forget to remove them.

Bdtrotte commented on WallsAndHoles/editor.cpp:
```diff
@@ -243,18 +237,17 @@ void Editor::setTileMap(TileMap *tileMap)
         mTileTemplateSetsView->setDefaultTileTemplateSet(nullptr);
 
-    delete mMap2Mesh;
+    if (mMap2Mesh != nullptr)
```
I think delete on a nullptr is safe

Bdtrotte commented on WallsAndHoles/map2mesh.cpp.orig:
```diff
@@ -0,0 +1,131 @@
+
```
STRAY FILE (Tima is a stat buffer)

Bdtrotte commented on WallsAndHoles/map2mesh.h.orig:
```diff
@@ -0,0 +1,104 @@
+#ifndef TIMAMAP2MESH_H
```
rm *.orig

Bdtrotte commented on WallsAndHoles/meshview.cpp:
```diff
@@ -69,188 +94,112 @@ void MeshView::activateTool(QString name) {
+    update();
 }
 
 void MeshView::load(QString path){
```
Just remove these

timoffex replied:
I can remove load(), but I can't remove save() yet because it is used by Editor::exportMapMesh().

timoffex replied:
I'll hold off on doing that though until I get time to work on a nice abstract save/load system.

Bdtrotte commented on WallsAndHoles/objtools.h:
```diff
@@ -1,156 +1,244 @@
 #ifndef OBJTOOLS_H
```
an implementation file would be nice. And probably wrap it in a namespace

Bdtrotte reviewed (changes requested):
Nice stuff.

A few rather trivial changes requested, and some stray files that need be removed.

Get it merged with master.

I still want to do some more testing in the actual app before merge (I was getting some strange behavior). and look through some of the code a bit more. Particularly, I'm pretty sure that when a tile changes, its mesh is just removed, and a new one is added, which is of course fine, but maybe there could be a special route if just the texture/material changes so all the data doesn't have to be changed, but just the uniforms it is rendered with. Maybe those changes could come in in the next commit that adds textures and materials to TileTemplates.

In any case, LOTS of work which is greatly appreciated. We should be able to merge by the end of today

timoffex committed:
Cleaned up AbstractRenderer::checkGLErrors()

timoffex committed:
Enabled back-face culling.

Note: Map2Mesh still outputs faces based on a “base” and “top” height,
where the base is the minimum of all the surrounding top heights.
Therefore, if a tile has at least one neighboring tile with a lower
height, it will draw all of its vertical faces with that height.

timoffex committed:
Cleaned up end of SimpleTexturedRenderer::createObjectBuffers()

timoffex committed:
Removed SimpleTexturedRenderer::cleanMaps()

timoffex committed:
removed stray .orig files

timoffex committed:
Tiny change in editor.cpp

timoffex committed:
Added makeCurrent() to MeshView before useGL() is called.
